indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorStateManager.java

```diff
@@ -0,0 +1,334 @@
+        Math.max(healthinessThreshold, unhealthinessThreshold),
+        "numExceptionEventsToStore must be greater than or equal to both "
+        + "healthinessThreshold and unhealthinessThreshold"
+    );
+    this.eventStore = new ExceptionEventStore(
+        supervisorConfig.getMaxStoredExceptionEvents(),
+        supervisorConfig.isStoringStackTraces()
+    );
+    this.completedTaskHistory = new CircularBuffer<>(Math.max(healthinessTaskThreshold, unhealthinessTaskThreshold));
+    this.stateHistory = new CircularBuffer<>(Math.max(healthinessThreshold, unhealthinessThreshold));
+  }
+
+  public Optional<State> setStateIfNoSuccessfulRunYet(State newState)
+  {
+    if (!atLeastOneSuccessfulRun) {
+      return Optional.of(setState(newState));
+    }
+    return Optional.absent();
+  }
+
+  public State setState(State newState)
+  {
+    if (newState.equals(State.SUSPENDED)) {
+      atLeastOneSuccessfulRun = false; // We want the startup states again after being suspended
+    }
+    this.supervisorState = newState;
+    return newState;
+  }
+
+  public void storeThrowableEvent(Throwable t)
+  {
+    if (t instanceof PossiblyTransientStreamException && atLeastOneSuccessfulRun) {
+      t = new TransientStreamException(t);
+    }
+
+    eventStore.storeThrowable(t);
+    currentRunSuccessful = false;
+  }
+
+  public void storeCompletedTaskState(TaskState state)
+  {
+    completedTaskHistory.add(state);
+  }
+
+  public void markRunFinishedAndEvaluateHealth()
+  {
+    if (currentRunSuccessful) {
+      atLeastOneSuccessfulRun = true;
+    }
+
+    State currentRunState = State.RUNNING;
+
+    for (Map.Entry<Class, Queue<ExceptionEvent>> events : eventStore.getNonTransientRecentEvents().entrySet()) {
+      if (events.getValue().size() >= unhealthinessThreshold) {
+        if (events.getKey().equals(NonTransientStreamException.class)) {
+          currentRunState = getHigherPriorityState(currentRunState, State.UNABLE_TO_CONNECT_TO_STREAM);
+        } else if (events.getKey().equals(TransientStreamException.class) ||
+                   events.getKey().equals(PossiblyTransientStreamException.class)) {
+          currentRunState = getHigherPriorityState(currentRunState, State.LOST_CONTACT_WITH_STREAM);
+        } else {
+          currentRunState = getHigherPriorityState(currentRunState, State.UNHEALTHY_SUPERVISOR);
+        }
+      }
+    }
+
+    // Evaluate task health
+    if (supervisorState == State.UNHEALTHY_TASKS) {
+      boolean tasksHealthy = completedTaskHistory.size() >= healthinessTaskThreshold;
+      for (int i = 0; i < Math.min(healthinessTaskThreshold, completedTaskHistory.size()); i++) {
+        if (completedTaskHistory.getLatest(i) != TaskState.SUCCESS) {
+          tasksHealthy = false;
+        }
+      }
+      if (tasksHealthy) {
+        currentRunState = State.RUNNING;
+      } else {
+        currentRunState = State.UNHEALTHY_TASKS;
+      }
+    } else {
```

Comment:
Since UNHEALTHY_SUPERVISOR is higher priority than UNHEALTHY_TASKS, I think you could skip this else block when currentRunState has been set to UNHEALTHY_SUPERVISOR